packages/contracts-bedrock/src/L2/GasTank.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import { IL2ToL2CrossDomainMessenger } from "interfaces/L2/IL2ToL2CrossDomainMessenger.sol";
+import { ICrossL2Inbox, Identifier } from "interfaces/L2/ICrossL2Inbox.sol";
+import { IGasTank } from "interfaces/L2/IGasTank.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { SafeSend } from "src/universal/SafeSend.sol";
+
+contract GasTank is IGasTank {
+    uint256 public constant MAX_DEPOSIT = 0.01 ether;
+    uint256 public constant WITHDRAWAL_DELAY = 7 days;
+    // TODO: Calculate claim overhead
+    uint256 public constant CLAIM_OVERHEAD = 100_000;
+
+    IL2ToL2CrossDomainMessenger public constant MESSENGER =
+        IL2ToL2CrossDomainMessenger(Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER);
+
+    mapping(address => uint256) public balanceOf;
+    mapping(address gasProvider => Withdrawal) public withdrawals;
+    mapping(bytes32 => bool) public claimed;
+    mapping(address gasProvider => mapping(bytes32 msgHash => bool)) public flaggedMessages;
+
+    // Deposit funds into the gas tank, from which the relayer can claim the repayment after relaying
+    function deposit(address _to) external payable {
+        uint256 newBalance = balanceOf[_to] + msg.value;
+
+        if (newBalance > MAX_DEPOSIT) revert MaxDepositExceeded();
+
+        balanceOf[_to] = newBalance;
+        emit Deposit(_to, msg.value);
+    }
+
+    function initiateWithdrawal(uint256 amount) external {
+        // Ensure the caller has enough balance
+        if (balanceOf[msg.sender] < amount) revert InsufficientBalance();
+
+        // Record the pending withdrawal
+        withdrawals[msg.sender] = Withdrawal({ timestamp: block.timestamp, amount: amount });
+
+        // Emit an event for the withdrawal initiation
+        emit WithdrawalInitiated(msg.sender, amount);
+    }
+
+    function finalizeWithdrawal(address to) external {
+        Withdrawal memory withdrawal = withdrawals[msg.sender];
+
+        // Ensure the withdraw is not pending
+        if (block.timestamp < withdrawal.timestamp + WITHDRAWAL_DELAY) revert WithdrawPending();
+
+        // Update the balance
+        uint256 amount = balanceOf[msg.sender] < withdrawal.amount ? balanceOf[msg.sender] : withdrawal.amount;
+        balanceOf[msg.sender] -= amount;
+
+        // Clear the pending withdrawal
+        delete withdrawals[msg.sender];
+
+        // Send the funds to the recipient
+        new SafeSend{ value: amount }(payable(to));
+
+        emit WithdrawalFinalized(msg.sender, to, amount);
+    }
+
+    // Flag a message into the gas tank so the relayer is aware of it, and can claim the funds after relaying
+    function flag(bytes32 rootMessageHash) external {
+        flaggedMessages[msg.sender][rootMessageHash] = true;
+        emit Flagged(rootMessageHash);
```

Comment:
what do you think of also adding the gas provider to this event? might help with the relayer, since it would be able to track which gas providers have flagged a given `rootMessageHash`